gnc/fam/src/fam.cc

```diff
@@ -67,20 +100,185 @@ void Fam::CtlCallBack(const ff_msgs::FamCommand & c) {
+  Eigen::Matrix<float, 12, 1> thrust_per_nozzle = forcetorque2thrust_ * forcetorque;
+  const int fam_nozzle_dirs[][4] = {{1, 2,  7,  8}, {3, 4,  9, 10}, {5, 6, 11, 12}};
+  // prevents negative thrust from being commanded by adding the min value below zero to the null vectors
+  // negative thrust occurs because the inverse matrix distributes thrust accross the
+  // nozzles aligned in an axis equally, even if they are pointed in opposite directions
+  for (int i = 0; i < 3; i++) {
+    float min_thrust = 0.0;
+    for (int j = 0; j < 4; j++) {
+      float t = thrust_per_nozzle(fam_nozzle_dirs[i][j] - 1);
+      if (t < min_thrust)
+        min_thrust = t;
+    }
+    if (min_thrust < 0) {
+      for (int j = 0; j < 4; j++)
+        thrust_per_nozzle(fam_nozzle_dirs[i][j] - 1) -= min_thrust;
+    }
+  }
+  pmc1_nozzle_thrusts = thrust_per_nozzle.head<6>();
+  pmc2_nozzle_thrusts = thrust_per_nozzle.tail<6>();
+}
+
+float Fam::Lookup(int table_size, const float lookup[], const float breakpoints[], float value) {
+  if (value < breakpoints[0])
+    return lookup[0];
+  if (value > breakpoints[table_size - 1])
+    return lookup[table_size - 1];
+  int start = 0, end = table_size;
+  while (start < end - 1) {
+    int mid = (start + end) / 2;
+    if (breakpoints[mid] <= value)
+      start = mid;
+    else
+      end = mid;
+  }
+
+  float ratio = (value - breakpoints[start]) / (breakpoints[start + 1] - breakpoints[start]);
+  return (1.0 - ratio) * lookup[start] + ratio * lookup[start + 1];
+}
+
+float Fam::ComputePlenumDeltaPressure(float impeller_speed, Eigen::Matrix<float, 6, 1> & discharge_coeff,
+                                     const Eigen::Matrix<float, 6, 1> & nozzle_thrusts) {
+  // lookup_data_tables
+  float total_thrust = (nozzle_thrusts.array() / discharge_coeff.array()).sum();
+  float cdp = Lookup(THRUST_LOOKUP_SIZE, THRUST_LOOKUP_CDP, THRUST_LOOKUP_BREAKPOINTS,
+                     total_thrust / (impeller_speed * impeller_speed));
+  const float units_in_2_m = 0.0254;
+  const float impeller_diameter = 5.5 * units_in_2_m;
+  const float air_density = 1.2;  //  [kg/m^3]   Air density inside of the ISS (some places in code this was 1.225?)
+  return impeller_speed * impeller_speed * impeller_diameter * impeller_diameter * air_density * cdp;
+}
+
+void Fam::CalcPMServoCmd(bool is_pmc1, uint8_t speed_gain_cmd, const Eigen::Matrix<float, 6, 1> & nozzle_thrusts,
+    uint8_t & impeller_speed_cmd, Eigen::Matrix<float, 6, 1> & servo_pwm_cmd,
+    Eigen::Matrix<float, 6, 1> & nozzle_theta_cmd,
+    Eigen::Matrix<float, 6, 1> & normalized_pressure_density, Eigen::Matrix<float, 6, 1> & command_area_per_nozzle) {
+  if (speed_gain_cmd > 3)
+    ROS_FATAL("Invalid speed command %d.", speed_gain_cmd);
+
+  Eigen::Matrix<float, 6, 1> discharge_coeff, nozzle_widths;
+  // Nozzle Discharge coefficient.  Determined via Test
+  if (is_pmc1)
+    discharge_coeff << 0.914971062, 0.755778254, 0.940762925, 0.792109779, 0.92401881 , 0.930319765;
```

Comment:
Ok, this is disturbing. I don't know of any reason why the left and right prop modules would have systematically different nozzle discharge coefficients across different Astrobee hardware units, so probably these numbers represent measurements from the two prop modules of a specific Astrobee hardware unit long ago, and they were semi-accidentally propagated for use on all Astrobee units. I notice the fourth entry in particular varies quite a lot between the two modules.

I think you did the right thing here in terms of exactly matching the output of the SIMULINK implementation so we don't break anything by switching. But we should also probably file an issue to figure out how to set these parameters more accurately. Maybe we don't have real discharge coefficients for all the hardware units, but if I understand this correctly, it would probably make more sense to at least use the average coefficients from the old calibration for both prop modules rather than blindly assuming the left and right modules will differ in the same way across all units. Jesse might have more insight.

Reply:
I was disturbed by this too. Agreed, let's stick to the old behavior for now.